go/def.bzl

```diff
@@ -891,57 +761,56 @@ def _cgo_codegen_impl(ctx):
       if lib.basename.startswith('lib') and lib.basename.endswith('.so'):
-        dirname = _short_path(lib)[:-len(lib.basename)]
-        linkopts += ['-L', dirname, '-l', lib.basename[3:-3]]
+        linkopts += ['-L', lib.dirname, '-l', lib.basename[3:-3]]
       else:
-        linkopts += [_short_path(lib)]
+        linkopts += [lib.path]
     linkopts += d.cc.link_flags
 
-  # collect files from $(SRCDIR), $(GENDIR) and $(BINDIR)
-  tree_layout = {}
-  for s in srcs:
-    tree_layout[s.path] = _short_path(s)
-
   p = _pkg_dir(ctx.label.workspace_root, ctx.label.package) + "/"
   if p == "./":
     p = "" # workaround when cgo_library in repository root
   out_dir = (ctx.configuration.genfiles_dir.path + '/' +
              p + ctx.attr.outdir)
   cc = ctx.fragments.cpp.compiler_executable
-  cmds = symlink_tree_commands(out_dir + "/src", tree_layout) + [
+  cmds = [
       'export GOROOT=$(pwd)/' + ctx.file.go_tool.dirname + '/..',
       # We cannot use env for CC because $(CC) on OSX is relative
       # and '../' does not work fine due to symlinks.
       'export CC=$(cd $(dirname {cc}); pwd)/$(basename {cc})'.format(cc=cc),
       'export CXX=$CC',
+      'objdir="%s/gen"' % out_dir,
       'execroot=$(pwd)',
-      'filter_tags="$execroot/%s"' % ctx.executable._filter_tags.path,
-      'objdir="$(pwd)/%s/gen"' % out_dir,
       'mkdir -p "$objdir"',
-      # The working directory must be the directory of the target go package
-      # to prevent cgo from prefixing mangled directory names to the output
-      # files.
-      # TODO(#350): this causes errors if sources are in multiple directories.
-      'cd "%s/src/$(dirname %s)"' % (out_dir, _short_path(ctx.files.srcs[0])),
-      'unfiltered_go_files=(%s)' % ' '.join(["'%s'" % f.basename for f in go_srcs]),
+      'unfiltered_go_files=(%s)' % ' '.join(["'%s'" % f.path for f in go_srcs]),
       'filtered_go_files=()',
       'for file in "${unfiltered_go_files[@]}"; do',
-      '  if "$filter_tags" -cgo -quiet "$file"; then',
+      '  name=$(basename "$file" .go)',
+      '  if %s -cgo -quiet "$file"; then' % ctx.executable._filter_tags.path,
       '    filtered_go_files+=("$file")',
       '  else',
-      '    grep --max-count 1 "^package " "$file" >"$objdir/$(basename "$file" .go).cgo1.go"',
-      '    echo -n >"$objdir/$(basename "$file" .go).cgo2.c"',
+      '    grep --max-count 1 "^package " "$file" >"$objdir/$name.go"',
+      '    echo -n >"$objdir/$name.c"',
       '  fi',
       'done',
       '"$GOROOT/bin/go" tool cgo -objdir "$objdir" -- %s "${filtered_go_files[@]}"' %
           ' '.join(['"%s"' % copt for copt in copts]),
-      'rm -f $objdir/_cgo_.o $objdir/_cgo_flags']
+      # Rename the outputs using glob so we don't have to understand cgo's mangling
+      # TODO(#350): might be fixed by this?.
+      'for file in "${filtered_go_files[@]}"; do',
+      '  name=$(basename "$file" .go)',
+      '  mv $objdir/*$name.cgo1.go $objdir/$name.go',
```

Comment:
I have made the change.
I think it could be properly fixed with $(find "$objdir" -maxdepth 1 -name "*$name.cgo.go") but that adds a dependency on a working version of find, so I decided not to.
The script is already not safe if there are spaces in the name (for instance, the rm at the bottom) so in the short term it's not a problem, and we plan to kill this for a go program in the next couple of weeks anyway.